chacha20/src/rng.rs

```diff
@@ -0,0 +1,1069 @@
+// muts when zeroize isn't enabled
+macro_rules! impl_zeroize_from {
+    ($from:ty, $struct:ident) => {
+        impl From<$from> for $struct {
+            #[cfg(feature = "zeroize")]
+            fn from(mut value: $from) -> Self {
+                let input = Self(value);
+                value.zeroize();
+                input
+            }
+            #[cfg(not(feature = "zeroize"))]
+            fn from(value: $from) -> Self {
+                Self(value)
+            }
+        }
+    };
+}
+
+// macro for ZeroizeOnDrop impl for wrappers
+macro_rules! impl_zeroize_on_drop {
+    ($struct:ident) => {
+        #[cfg(feature = "zeroize")]
+        impl Drop for $struct {
+            fn drop(&mut self) {
+                self.0.zeroize();
+            }
+        }
+        #[cfg(feature = "zeroize")]
+        impl ZeroizeOnDrop for $struct {}
+    };
+}
+
+/// The seed for ChaCha20. Implements ZeroizeOnDrop when the
+/// zeroize feature is enabled.
+#[derive(PartialEq, Eq)]
+#[cfg_attr(feature = "serde1", derive(Serialize, Deserialize))]
+pub struct Seed([u8; 32]);
+
+impl Default for Seed {
+    fn default() -> Self {
+        Self([0u8; 32])
+    }
+}
+impl AsRef<[u8; 32]> for Seed {
+    fn as_ref(&self) -> &[u8; 32] {
+        &self.0
+    }
+}
+impl AsMut<[u8]> for Seed {
+    fn as_mut(&mut self) -> &mut [u8] {
+        self.0.as_mut()
+    }
+}
+
+impl_zeroize_from!([u8; 32], Seed);
+
+impl Debug for Seed {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl_zeroize_on_drop!(Seed);
+
+/// An internally used trait to help with zeroizing unsigned ints that are primarily
+/// used for le bytes
+trait ZeroizeToLeBytes {
+    type Output;
+    fn zeroize_to_le_bytes(&mut self) -> Self::Output;
+}
+
+// Define macro for ensuring zeroization of an unsigned int before it is converted to
+// le bytes
+#[cfg(feature = "zeroize")]
+macro_rules! impl_zeroize_to_le_bytes {
+    ($type:ident, $output_size:expr, $byte_wrapper:ident) => {
+        struct $byte_wrapper([u8; $output_size]);
+
+        impl_zeroize_on_drop!($byte_wrapper);
+
+        impl ZeroizeToLeBytes for $type {
+            type Output = $byte_wrapper;
+            fn zeroize_to_le_bytes(&mut self) -> $byte_wrapper {
+                let bytes = $byte_wrapper(self.to_le_bytes());
+                self.zeroize();
+                bytes
+            }
+        }
+    };
+}
+#[cfg(feature = "zeroize")]
+impl_zeroize_to_le_bytes!(u64, 8, Zeroizing8Bytes);
+#[cfg(feature = "zeroize")]
+impl_zeroize_to_le_bytes!(u128, 16, Zeroizing16Bytes);
+
+/// A zeroizable wrapper for set_word_pos() input that can be assembled from:
+/// * `u64`
+/// * `[u8; 5]`
+///
+/// There would be a minor performance benefit from using a `[u8; 5]`, as it
+/// avoids some copies, bit operations, and extra zeroizing.
+pub struct WordPosInput([u8; 5]);
+
+impl_zeroize_from!([u8; 5], WordPosInput);
```

Comment:
As a general note regarding zeroization: it's probably best reserved for values which are themselves cryptographic keys or derived from the key, such as the internal cipher state. (Note that this doesn't include the keystream, where the cipher's diffusion prevents it from being reversed back into the key/cipher state)

I don't see the value in zeroization here, and generally using it on anything transient is questionable.

Reply:
Gotcha. Since there are some `rust-crypto` libraries that are able to use objects that `impl CryptoRngCore` to generate private keys, I suppose the internal buffers might still need zeroizing?

Reply:
I think the general rule there is it's up to the caller to decide if that output represents a long-term secret, and if so, zeroization is up to them

Reply:
Alright. Will add the `zeroize` method. Good call.

Reply:
I couldn't find a way to zeroize the `BlockRng`'s `results` member (without using `ZeroizeOnDrop`), so I went ahead and replaced both `BlockRngCore` and `BlockRng` dependent code. Most of the code was already in `rng.rs`, but the rest came from `rand_core`.